### PR TITLE
[Build] Make cmark build a build-script product

### PR DESCRIPTION
**Pull request by drexin**

### CODE REVIEW

File: utils/swift_build_support/tests/products/test_cmark.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        shutil.rmtree(self.workspace.build_root)
+        shutil.rmtree(self.workspace.source_root)
+        sys.stdout = self._orig_stdout
+        sys.stderr = self._orig_stderr
+        shell.dry_run = False
+        self.workspace = None
+        self.toolchain = None
+        self.args = None
+
+    def test_build(self):
```
Comment (drexin) [inline utils/swift_build_support/tests/products/test_cmark.py:82]: Need to fill in the actual test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

File: utils/swift_build_support/swift_build_support/products/cmake_product.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        if self.args.reconfigure or not os.path.isfile(cmake_cache_path) or \
+                (generator_output_path and not os.path.isfile(generator_output_path)):
+            if not os.path.exists(self.build_dir):
+                os.makedirs(self.build_dir)
+
+            # Use `cmake-file-api` in case it is available.
+            query_dir = os.path.join(self.build_dir, ".cmake", "api", "v1", "query")
+            if not os.path.exists(query_dir):
+                os.makedirs(query_dir)
+            open(os.path.join(query_dir, "codemodel-v2"), 'a').close()
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/products/cmake_product.py:69]: Is this really the canonical way to do this in python? I guess it works.

File: utils/swift_build_support/swift_build_support/products/cmake_product.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+import os
+
+from . import product
+from .. import cmake
+from .. import shell
+
+
+class CMakeProduct(product.Product):
+    def should_include_host_in_lipo(self, host_target):
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/products/cmake_product.py:21]: I think this method should be in the parent product class. Seems like something that might be useful for all products.

File: utils/swift_build_support/swift_build_support/products/cmake_product.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        if self.args.cross_compile_hosts:
+            if host_target.startswith("macosx") or \
+               host_target.startswith("iphone") or \
+               host_target.startswith("appletv") or \
+               host_target.startswith("watch"):
+                return True
+        return False
+
+    def host_install_destdir(self, host_target):
+        if self.args.cross_compile_hosts:
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/products/cmake_product.py:31]: This method also I think should be in the parent product class. Seems like something that might be useful for all products.

**Review by gottesmm — APPROVED** [review]
Small bit of quick feedback. I am probably going to move those two methods to a parent class in a forthcoming commit. Looks good over all though!

**Comment by drexin** [comment]
@swift-ci smoke test

File: less [hunk 5]
```diff
@@ -0,0 +1,44 @@
+note: skipping Xcode version check
+[/Users/drexin/src/swift-source/swift/utils/build-script] NOTE: Using toolchain default
++ mkdir -p /Users/drexin/src/swift-source/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/validation-test-macosx-x86_64/BuildSystem/Output/infer_dumps_deps_if_verbose_build.test.tmp/Ninja-DebugAssert
+-- Build Graph Inference --
+Initial Product List:
+    llvm
+    swift
+    swiftpm
+Final Build Order:
```
Comment (gottesmm) [inline less:9]: @drexin I don't think you meant to commit this one.

Comment (drexin) [reply less:9]: dang it... thanks

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test macos

**Comment by drexin** [comment]
@swift-ci smoke test windows

**Comment by drexin** [comment]
@swift-ci smoke test windows

**Comment by benlangmuir** [comment]
This is failing: https://ci.swift.org/job/oss-swift-package-macos/5874/console

```
17:24:52 CMake Error at cmake/modules/SwiftSharedCMakeConfig.cmake:200 (include):
17:24:52   include could not find load file:
17:24:52 
17:24:52     /Users/buildnode/jenkins/workspace/oss-swift-package-macos/build/buildbot_osx/cmark-macosx-arm64/src/cmarkTargets.cmake
17:24:52 Call Stack (most recent call first):
17:24:52   cmake/modules/SwiftSharedCMakeConfig.cmake:214 (swift_common_standalone_build_config_cmark)
17:24:52   CMakeLists.txt:521 (swift_common_standalone_build_config)
```

I'm going to revert it https://github.com/apple/swift/pull/37604.  Please test with "swift-ci please build toolchain" before re-landing.

**Pull request merged**